### PR TITLE
Revert "Reland 3rd try [finishing colesbury's PR 100642] Guard on nn.Module dicts and type (#109323)" + Forward fixes + test

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110964

This reverts commit f786fbdebdd24d3a6807e3b9fbf055836db4ad60.

Forward fixes

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_recompiles.py [hunk 1]
```diff
@@ -295,6 +295,27 @@ def run_dynamic_comp_dynamic_param():
...
+        model = SimpleDropout()
+        x = torch.randn(10)
+        counter = torch._dynamo.testing.CompileCounter()
+        model = torch.compile(model, backend=counter, fullgraph=True)
+        for _ in range(20):
+            model.eval()
+            model(x)
+            model.train()
+            model(x)
+        self.assertEqual(counter.frame_count, 2)
```
Comment (voznesenskym) [inline test/dynamo/test_recompiles.py:317]: This test fails pre revert with 8 frames.

File: torch/csrc/dynamo/guards.cpp [hunk 2]
```diff
@@ -579,180 +578,13 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
...
+// NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
 static PyMethodDef _methods[] = {
-    {"check_type_id", check_type_id, METH_VARARGS, NULL},
-    {"check_obj_id", check_obj_id, METH_VARARGS, NULL},
-    {"assert_size_stride", assert_size_stride, METH_VARARGS, NULL},
-    {"nn_module_guard", nn_module_guard, METH_O, NULL},
+    {"check_type_id", check_type_id, METH_VARARGS, nullptr},
+    {"check_obj_id", check_obj_id, METH_VARARGS, nullptr},
+    {"assert_size_stride", assert_size_stride, METH_VARARGS, nullptr},
     {"dict_version", dict_version, METH_VARARGS, NULL},
```
Comment (voznesenskym) [inline torch/csrc/dynamo/guards.cpp:586]: This line should have been reverted - but we need it for pytree tracing https://github.com/pytorch/pytorch/pull/108533 - this is the only forward fix.

**Review by ezyang — APPROVED** [review]

**Review by anijain2305 — APPROVED** [review]

**Pull request closed without merging**